File: src/geonhyeong/week4/베스트앨범.swift
/*
*  베스트 앨범.swift
*  Question Link: https://school.programmers.co.kr/learn/courses/30/lessons/42579
*  Primary idea:   <key-value>
*                   1. 
*
*  걸린 시간 : 05:00, 53:00
*  Time Complexity : O()
*  Space Complexity : O()
*  Runtime:  ms
*  Memory Usage:  MB
*/

func solution(_ genres:[String], _ plays:[Int]) -> [Int] {
  var plays = plays
  var dic = [String:[Int]]()
  var result = [Int]()
  
  for (i, genre) in genres.enumerated() {
    dic[genre, default: []].append(plays[i])
  }
  
  // 장르 내에서 많이 재생된 노래
  let dic2 = dic.mapValues{$0.sorted(by:>)} // ["pop": [2500, 600], "classic": [800, 500, 150]]
  // 속한 노래가 많이 재생된 장르
  let dic3 = dic2.sorted(by: {$0.value.reduce(0, +) > $1.value.reduce(0, +)}).map{$0.key} // ["pop", "classic"]
  
  for genre in dic3 {
    var cnt = 0
    for value in dic2[genre]! {
      if cnt == 2 { // 최대 두 개까지
        continue
      }
      
      let index = plays.firstIndex(of: value)!
      result.append(index)
      plays[index] = 0
      cnt += 1
    }
  }
  return result
}

print(solution(["classic", "pop", "classic", "classic", "pop"]  , [500, 600, 150, 800, 2500]  )) // [4, 1, 3, 0]

File: src/taeeun/week4/베스트앨범.js
function solution(genres, plays) {
    var answer = [];
    const music=[];
    const genreMap=new Map();

    for (let i=0; i<genres.length; i++) {
        music.push({genre:genres[i], plays:plays[i], index: i});
        if (!genreMap.has(genres[i])) genreMap.set(genres[i], plays[i]);
        else genreMap.set(genres[i], genreMap.get(genres[i])+plays[i]);
    }
    
    const genreArray=[...genreMap];
    genreArray.sort((a,b)=> b[1]-a[1]);
    music.sort((a,b)=> {
        if (genreMap.get(a.genre)!==genreMap.get(b.genre)) return genreMap.get(b.genre)-genreMap.get(a.genre)
        if (a.plays!==b.plays) return b.plays-a.plays;
        return a.index-b.index;
    })

    for (const [g, _] of genreArray) {
        let cnt=0;
        for (const {genre, plays, index} of music) {
            if (genre===g && cnt<2) {
                answer.push(index);
                cnt++;
            }
        }
    }

    return answer;
}
